### DISCUSSION TITLE
Software Diagrams - Plant UML vs Mermaid | Dan Does Code

### DISCUSSION THREAD

**Discussion by giscus-migrator** [danielmackay/nextjs-personal-blog#74] [Blog Comments]
Software Diagrams - Plant UML vs Mermaid
        
[https://www.dandoescode.com/blog/plantuml-vs-mermaid/](https://www.dandoescode.com/blog/plantuml-vs-mermaid/)

**Comment by giscus-migrator** [comment]
```header
Originally posted by RDMueller on 2022-11-06 01:06
```

Great comparison!  
Three comments:  
Have you already found [kroki.io?](http://kroki.io?) It is a server which supports plantuml, mermaid and many more diagram types.  
Regarding having to store the generated images - I am only aware of rendering engines (github, gitlab, asciidoc etc.) which generate the images on the fly. No need to store them with version control.  
Btw: are you aware that plantuml stores the sources of an image within the meta data of the png and svg. This is sometimes good to know.

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2022-11-06 08:18
```

Thanks for the feedback. I hadn't heard of [kroki.io](http://kroki.io). Cheers for the suggestion.  
Looks like it's an aggregator that routes the diagram code to the right rendering engine under the hood?

RE: stored images, correct this is only needed for PlantUML in Azure DevOps/GitHub as Mermaid is automatically rendered (which is a big winner for me).

I was aware of the SVG trick, but didn't know it worked for PNG also. Cheers!

**Reply by giscus-migrator** [reply]
```header
Originally posted by ArrozConCostra on 2022-12-20 05:30
```

didn't knwo about kroki neither, thanks!

**Comment by giscus-migrator** [comment]
```header
Originally posted by Christopher Fuhrman on 2022-11-20 02:56
```

Awesome comparison. It's important to know that [https://app.diagrams.net/](https://app.diagrams.net/) lets you insert PlantUML and Mermaid. The fact that you can also save the whole diagram as "blah.drawio.svg" (embedding the [diagrams.net](http://diagrams.net) source and PlantUML/Mermaid source) is very meta and cool. The only thing that doesn't quite work is the non-supported SVG sometimes, but this workaround is reasonable: [https://www.diagrams.net/do...](https://www.diagrams.net/doc/faq/svg-export-text-problems)

**Comment by giscus-migrator** [comment]
```header
Originally posted by Christopher Fuhrman on 2022-12-20 06:12
```

\> With PlantUML, you need to also check in the images to source control and these could easily become out of date. Winner: Mermaid

Not entirely accurate. See [https://stackoverflow.com/a...](https://stackoverflow.com/a/32771815) for some markup magic that works if your repo isn't private (uses the PlantUML web service to render the images inline).

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2022-12-20 17:17
```

Awesome! I wasn't aware of this. Thanks a lot for sharing!

**Comment by omac777** [comment]
I've actually tried both plantUML and mermaid.

Both in my opinion lack the most important kind of diagram which is the UML collaboration/communication diagram displaying the temporal/relationship at the same time within one diagram.  It is the perfect diagram for explaining contexts, how they connect and when they communicate.  It is the one that helps to crystallize what is elicited and then with the diagram you can get clarification and correct/tweak it to represent an existing real-world situation more accurately and then move on to the next step which is what you really want the scenario to be improved on.  It helps to communicate the change requests and helps everybody all stakeholders to participate be on the same page/wavelength for what is the system and for what everybody wants to system to morph into.

**Reply by omac777** [reply]
Rational Rose and Visual Paradigm offer the capability of making collaboration/communication diagrams.

I think plantUML/Mermaid and structurizr.com all fall short in capability without this type of diagram.
The structurizr.com Dynamic Diagram is inadequate.

**Comment by vibrog** [comment]
You wrote that GitGraph is not supported by PlantUML, but there are now [experimental support for generating git branching diagrams](https://forum.plantuml.net/5184/generate-git-graph). There are [examples of using component diagrams and sequence diagrams](https://drawson.medium.com/git-branching-diagrams-using-plantuml-abaa99696a32) for it. PlantUML also supports [Extended Backus-Naur Form (EBNF) diagrams](https://plantuml.com/ebnf), which can be used to draw branching and merging operations.

**Comment by RattanOnly** [comment]
Just read this comparison between PlantUML and Mermaid—super helpful! I’ve been using Mermaid for flowcharts but had no idea about the <a href="https://mermaidviewer.com/">mermaidviewer tutorial</a> options in VS Code. Makes things way easier than manually tweaking boxes. Both tools seem solid, though Mermaid’s setup is definitely quicker. Might stick with it for smaller projects!

**Comment by Beautyod** [comment]
Thanks to you, Daniel, for your good article (-:

**Comment by olek3** [comment]
I would use PlantUML features to draw. the sequence would be like

```plantuml
@startuml

title Sequence Diagram

autonumber

actor Customer as C
boundary "Web Server" as WS
entity "Warehouse" as WH
entity "Stripe" as S

C -> WS ++ : Submit Order

WS ->> WH: Check Stock
WS ->> S: Process Payment
WS ->> WH: Dispatch Order
return: Show Success


@enduml
```

**Comment by olek3** [comment]
> For State diagrams, syntax is also similar, but Mermaid renders the images slightly better. Not sure why PlantUML needs to have a line in the middle of each node. Winner: Mermaid

because that is the syntax of UML for state diagrams

```plantuml
@startuml
'https://plantuml.com/en-dark/state-diagram

'========================== форматирование ==================================

!$branch="main"
!$libPath="https://raw.githubusercontent.com/olek3/plantuml-includes/refs/heads/" + $branch + "/iuml"
!include $libPath/common.iuml
!include $libPath/state.iuml

'========================== статусы ==================================

title report task state diagram

state CREATED

state IN_PROGRESS {

    state CAPTURED

    state REPORT_REQUESTED

    state REPORT_DOWNLOADED
}

state REPORT_SENT : entry /save completed


state FAILED


[*] -> CREATED : Task created
CREATED --> CAPTURED : worker took the task /save dequeued, worker_id, captured_by
IN_PROGRESS --> CREATED : task released /task.run_count++

CAPTURED -> REPORT_REQUESTED : report requested 

REPORT_REQUESTED -> REPORT_DOWNLOADED : report received

REPORT_DOWNLOADED --> REPORT_SENT : report sent


IN_PROGRESS --> FAILED : error [attempt limit exceeded]

CAPTURED <- REPORT_REQUESTED : error receiving report /task.run_count++
REPORT_REQUESTED <- REPORT_DOWNLOADED : error sending report /task.run_count++

FAILED --> [*]

REPORT_SENT --> [*]



@enduml
```
